### ISSUE TITLE
Improved parallelization architecture in STELLOPT

### ISSUE THREAD

**Issue by lazersos** [PrincetonUniversity/STELLOPT#61]
The parallelization in STELLOPT should be improved to better handle users who wish to set `NOPTIMIZERS<NNODES` and `NOPTIMIZERS>NNODES`.  We must still respect the granularity of computational architecture but this should be possible.  There are three cases we need to address:

1. The user wants a given function evaluation spread across multiple nodes (`NOPTIMIZERS<NNODES`).

   - Check to make sure `NNODES` is evenly divisible by `NOPTIMIZERS` and adjust if necessary.
   - The interface to `DIAGNO`, `BEAMS3D`, and `FIELDLINES` needs to verify that the `shared` communicator is set based on the passed communicator (but this should be true).

2. The user wants a given node to be subdivided for multiple function evaluations (`NOPTIMIZERS>NNODES`).
   - Check to make sure that the threads on a given node are evenly divisible by `NOPTIMIZERS` and adjust if necessary.
   - Adjust `virtual_casing_mod.f90`, `stell_tools.f90`, and `mgrid_field_mod.f90` to not expect a shared communicator (remove `MPI_COMM_SPLIT` from routines).  Maybe add a check in case somebody codes up something weird.
   - Adjust logic in `DIAGNO`, `BEAMS3D`, and `FIELDLINES` to handle the allocation of shared_memory communicators in this scenario.

It may be valuable to add a routine to libstell which given a communicator assists in the determination of it's shared_memory properties.  For example, something like MPI_IS_SHARED(COMM,ier).  That way if the communicator spans a shared memory space we don't have to divide it, otherwise we subdivide it using split.

**Comment by lazersos** [comment]
Looking through the documentation of [MPI_COMM_SPLIT_TYPE](https://www.open-mpi.org/doc/v3.0/man3/MPI_Comm_split_type.3.php), it appears we don't need to do anything too special. I need to check this but if we pass a communicator to `MPI_COMM_SPLIT_TYPE` which is only part of a shared memory group then it acts like `MPI_COMM_DUP`.

**Comment by lazersos** [comment]
I seem to have at least the parallelization breakdown portion of the code working properly.  Pull from branch `Lazerson` to test.

**Comment by lazersos** [comment]
Testing last night. I can confirm the code behaves as expected and VMEC runs in fixed boundary.

**Comment by lazersos** [comment]
I did some more updates.  There was a slight modification needed to VMEC but now the code should be ready for full testing.  Please pull, and if possible could someone test STELLOPT with free boundary VMEC equilibria?

**Comment by lazersos** [comment]
This appears to be working in free boundary.  I'm going to close this issue and create a pull request for `Lazerson` into `develop`.

**Comment by landreman** [comment]
Sam- Thanks for coding this up. I did some tests and it mostly works for me, aside from these issues:

1. If a user for some reason request `noptimizers` larger than the number of mpi processes, `common_factor` calls itself recursively forever and the code hangs. It would be good to handle this case more gracefully.

2. If the user for some reason requests `noptimizers=0`, the code crashes with a divide-by-0 in line 69:

> IF (MOD(ngshar,noptimizers)/=0) THEN ! we need to redefine NOPTIMIZERS 

3. If you have several optimizer groups on a node, with multiple processes per group, the groups are made up of ranks that are not contiguous. This behavior is determined by line 118:

> color = MOD(myworkid,k)

There may be a performance advantage to having each optimizer group composed of adjacent ranks, which presumably are closer to each other on the chip. I'm not sure if this makes any significant difference in practice or not. To make the ranks adjacent, I believe the replacement line would be

> color = myworkid / (nshar / k)

For testing, I used the branch `Matt_mpi`, in which a file `stellopt_mpi` is written with detailed MPI information such as each processor's rank in various communicators. The 3 issues above are addressed in this branch. We could merge in that branch, or not, if you prefer to do things differently.

If anyone is interested, I found it useful to have a script that runs stellopt for many different values of (# of nodes), (# of processors per node), and noptimizers. The script is available at
[https://terpconnect.umd.edu/~mattland/stelloptMPIScan](https://terpconnect.umd.edu/~mattland/stelloptMPIScan)
